File: ffmpeg-service/app/voices.py
"""Catalogue des voix et des styles de lecture."""

from dataclasses import dataclass


@dataclass(frozen=True)
class Voice:
    id: str
    label: str
    gender: str  # "female" | "male"


# Voix natives de Gemini TTS (toutes multilingues, françaises comprises).
GEMINI_VOICES: tuple[Voice, ...] = (
    Voice("Kore", "Kore — ferme", "female"),
    Voice("Aoede", "Aoede — légère", "female"),
    Voice("Leda", "Leda — jeune", "female"),
    Voice("Zephyr", "Zephyr — lumineuse", "female"),
    Voice("Callirrhoe", "Callirrhoe — décontractée", "female"),
    Voice("Autonoe", "Autonoe — lumineuse", "female"),
    Voice("Despina", "Despina — douce", "female"),
    Voice("Erinome", "Erinome — claire", "female"),
    Voice("Laomedeia", "Laomedeia — enjouée", "female"),
    Voice("Achernar", "Achernar — tendre", "female"),
    Voice("Gacrux", "Gacrux — mûre", "female"),
    Voice("Pulcherrima", "Pulcherrima — affirmée", "female"),
    Voice("Vindemiatrix", "Vindemiatrix — délicate", "female"),
    Voice("Sulafat", "Sulafat — chaleureuse", "female"),
    Voice("Charon", "Charon — informative", "male"),
    Voice("Puck", "Puck — enjouée", "male"),
    Voice("Fenrir", "Fenrir — enthousiaste", "male"),
    Voice("Orus", "Orus — ferme", "male"),
    Voice("Enceladus", "Enceladus — soufflée", "male"),
    Voice("Iapetus", "Iapetus — claire", "male"),
    Voice("Umbriel", "Umbriel — décontractée", "male"),
    Voice("Algieba", "Algieba — veloutée", "male"),
    Voice("Algenib", "Algenib — rocailleuse", "male"),
    Voice("Rasalgethi", "Rasalgethi — informative", "male"),
    Voice("Alnilam", "Alnilam — ferme", "male"),
    Voice("Schedar", "Schedar — posée", "male"),
    Voice("Achird", "Achird — amicale", "male"),
    Voice("Zubenelgenubi", "Zubenelgenubi — naturelle", "male"),
    Voice("Sadachbia", "Sadachbia — vive", "male"),
    Voice("Sadaltager", "Sadaltager — experte", "male"),
)

EDGE_VOICES: tuple[Voice, ...] = (
    Voice("fr-FR-VivienneMultilingualNeural", "Vivienne", "female"),
    Voice("fr-FR-DeniseNeural", "Denise", "female"),
    Voice("fr-FR-EloiseNeural", "Eloise", "female"),
    Voice("fr-FR-RemyMultilingualNeural", "Rémy", "male"),
    Voice("fr-FR-HenriNeural", "Henri", "male"),
)

# Consignes de lecture ajoutées au texte envoyé à Gemini.
STYLES: dict[str, tuple[str, str]] = {
    "neutral": ("Neutre", ""),
    "dynamic": (
        "Dynamique",
        "Lis ce texte en français sur un ton dynamique et enthousiaste, avec un rythme "
        "entraînant, comme une vidéo courte sur les réseaux sociaux",
    ),
    "warm": (
        "Chaleureux",
        "Lis ce texte en français sur un ton chaleureux, souriant et proche, comme si tu conseillais un ami",
    ),
    "calm": (
        "Calme",
        "Lis ce texte en français sur un ton calme, posé et rassurant, sans te presser",
    ),
    "promo": (
        "Promo",
        "Lis ce texte en français comme une annonce promotionnelle énergique, en "
        "insistant sur les offres et les prix",
    ),
}

_GEMINI_BY_ID = {v.id.lower(): v for v in GEMINI_VOICES}
_EDGE_BY_ID = {v.id: v for v in EDGE_VOICES}


def resolve_gemini_voice(voice: str | None) -> Voice:
    """Voix Gemini à utiliser.

    Accepte aussi les anciens identifiants « fr-FR-Standard-X » encore stockés
    dans des jobs (B/D = voix masculine).
    """
    if voice and voice.lower() in _GEMINI_BY_ID:
        return _GEMINI_BY_ID[voice.lower()]
    if voice and voice.endswith(("-B", "-D")) or voice == "male":
        return _GEMINI_BY_ID["charon"]
    return _GEMINI_BY_ID["kore"]


def voice_gender(voice: str | None) -> str:
    """Genre d'une voix, quel que soit le moteur d'origine."""
    if not voice:
        return "female"
    if voice.lower() in _GEMINI_BY_ID:
        return _GEMINI_BY_ID[voice.lower()].gender
    if voice in _EDGE_BY_ID:
        return _EDGE_BY_ID[voice].gender
    if voice == "male" or voice.endswith(("-B", "-D")):
        return "male"
    return "male" if any(name in voice for name in ("Remy", "Henri", "Paul")) else "female"


def resolve_edge_voice(voice: str | None) -> Voice:
    """Voix Edge à utiliser ; une voix Gemini est remplacée par une voix Edge du même genre."""
    if voice in _EDGE_BY_ID:
        return _EDGE_BY_ID[voice]
    gender = voice_gender(voice)
    return next(v for v in EDGE_VOICES if v.gender == gender)


def edge_fallbacks(primary: Voice) -> list[Voice]:
    """Voix françaises de secours du même genre (jamais d'anglais)."""
    same = [v for v in EDGE_VOICES if v.gender == primary.gender and v != primary]
    return [primary, *same]


def style_instruction(style: str | None) -> str:
    return STYLES.get(style or "neutral", STYLES["neutral"])[1]


def catalog() -> dict:
    return {
        "gemini": [v.__dict__ for v in GEMINI_VOICES],
        "edge": [v.__dict__ for v in EDGE_VOICES],
        "styles": [{"id": k, "label": v[0]} for k, v in STYLES.items()],
    }

File: shared/voices.ts
/**
 * Voix et styles de lecture proposés pour les Reels.
 * Même catalogue que le service Python (ffmpeg-service/app/voices.py).
 */

export const TTS_ENGINES = ["gemini", "edge"] as const;
export type TtsEngine = (typeof TTS_ENGINES)[number];

export const TTS_STYLES = ["neutral", "dynamic", "warm", "calm", "promo"] as const;
export type TtsStyle = (typeof TTS_STYLES)[number];

export interface VoiceOption {
  id: string;
  label: string;
  gender: "female" | "male";
}

export const TTS_STYLE_OPTIONS: { id: TtsStyle; label: string; description: string }[] = [
  { id: "dynamic", label: "Dynamique", description: "Enthousiaste, rythme entraînant" },
  { id: "warm", label: "Chaleureux", description: "Souriant et proche" },
  { id: "promo", label: "Promo", description: "Annonce énergique des offres" },
  { id: "calm", label: "Calme", description: "Posé et rassurant" },
  { id: "neutral", label: "Neutre", description: "Lecture sans consigne" },
];

export const GEMINI_VOICES: VoiceOption[] = [
  { id: "Kore", label: "Kore — ferme", gender: "female" },
  { id: "Aoede", label: "Aoede — légère", gender: "female" },
  { id: "Leda", label: "Leda — jeune", gender: "female" },
  { id: "Zephyr", label: "Zephyr — lumineuse", gender: "female" },
  { id: "Callirrhoe", label: "Callirrhoe — décontractée", gender: "female" },
  { id: "Autonoe", label: "Autonoe — lumineuse", gender: "female" },
  { id: "Despina", label: "Despina — douce", gender: "female" },
  { id: "Erinome", label: "Erinome — claire", gender: "female" },
  { id: "Laomedeia", label: "Laomedeia — enjouée", gender: "female" },
  { id: "Achernar", label: "Achernar — tendre", gender: "female" },
  { id: "Gacrux", label: "Gacrux — mûre", gender: "female" },
  { id: "Pulcherrima", label: "Pulcherrima — affirmée", gender: "female" },
  { id: "Vindemiatrix", label: "Vindemiatrix — délicate", gender: "female" },
  { id: "Sulafat", label: "Sulafat — chaleureuse", gender: "female" },
  { id: "Charon", label: "Charon — informative", gender: "male" },
  { id: "Puck", label: "Puck — enjouée", gender: "male" },
  { id: "Fenrir", label: "Fenrir — enthousiaste", gender: "male" },
  { id: "Orus", label: "Orus — ferme", gender: "male" },
  { id: "Enceladus", label: "Enceladus — soufflée", gender: "male" },
  { id: "Iapetus", label: "Iapetus — claire", gender: "male" },
  { id: "Umbriel", label: "Umbriel — décontractée", gender: "male" },
  { id: "Algieba", label: "Algieba — veloutée", gender: "male" },
  { id: "Algenib", label: "Algenib — rocailleuse", gender: "male" },
  { id: "Rasalgethi", label: "Rasalgethi — informative", gender: "male" },
  { id: "Alnilam", label: "Alnilam — ferme", gender: "male" },
  { id: "Schedar", label: "Schedar — posée", gender: "male" },
  { id: "Achird", label: "Achird — amicale", gender: "male" },
  { id: "Zubenelgenubi", label: "Zubenelgenubi — naturelle", gender: "male" },
  { id: "Sadachbia", label: "Sadachbia — vive", gender: "male" },
  { id: "Sadaltager", label: "Sadaltager — experte", gender: "male" },
];

export const EDGE_VOICES: VoiceOption[] = [
  { id: "fr-FR-VivienneMultilingualNeural", label: "Vivienne", gender: "female" },
  { id: "fr-FR-DeniseNeural", label: "Denise", gender: "female" },
  { id: "fr-FR-EloiseNeural", label: "Eloise", gender: "female" },
  { id: "fr-FR-RemyMultilingualNeural", label: "Rémy", gender: "male" },
  { id: "fr-FR-HenriNeural", label: "Henri", gender: "male" },
];

export const DEFAULT_VOICE: Record<TtsEngine, string> = {
  gemini: "Kore",
  edge: "fr-FR-VivienneMultilingualNeural",
};

export const DEFAULT_TTS_STYLE: TtsStyle = "dynamic";

export function voicesFor(engine: TtsEngine): VoiceOption[] {
  return engine === "gemini" ? GEMINI_VOICES : EDGE_VOICES;
}
